dmriprep/utils/vectors.py

```diff
@@ -0,0 +1,295 @@
+
+    # Check length equality
+    # Correct b0's in bvecs mistated as 10's.
+    bvecs[np.where(np.any(abs(bvecs) >= 10, axis=1) == True)] = [1, 0, 0]
+
+    # Ensure that zeros are zero floats.
+    bvecs[np.where(bvals == 0)] = float(0)
+
+    # Check for bval-bvec discrepancy
+    if len(bvecs[np.where(np.logical_and(bvals > 50, np.all(abs(bvecs) == np.array([0, 0, 0]), axis=1)))]) > 0:
+        raise ValueError("Corrupted bval/bvecs detected. Check to ensure volumes with a diffusion weighting are not listed as B0s in the bvecs.")
+
+    # Raise warnings if sampling scheme is not eddy-friendly
+    if ((len(bvals) < 10) and (max(bvals) < 1500)) or ((len(bvals) < 30) and (max(bvals) < 5000)):
+        raise UserWarning('Too few directions in this data. Use of eddy is not recommended.')
+
+    # Save corrected
+    np.savetxt(fbval, bvals.T.astype('int'), fmt='%d', newline=' ')
+    np.savetxt(fbvec, bvecs.T.astype('float'), fmt='%.8f')
+
+    # Rescale bvals
+    fbval_rescaled = "%s%s" % (vector_dir, "/bval_scaled.bval")
+    bval_rescaled = rescale_bval(bvals.astype('float'), bvecs.astype('float'))
+    # Get b order of magnitude and round bvals by it
+    bmag = int(np.log10(np.max(bval_rescaled))) - 1
+    bval_rescaled_mag = round_bvals(bval_rescaled, bmag=bmag)
+    np.savetxt(fbval_rescaled, bval_rescaled_mag.T.astype('int'), fmt='%d', newline=' ')
+
+    # Rescale bvecs
+    fbvec_rescaled = "%s%s" % (vector_dir, "/bvec_scaled.bvec")
+    bvec_rescaled = rescale_bvec(bvecs)
+    np.savetxt(fbvec_rescaled, bvec_rescaled.T.astype('float'), fmt='%.8f')
+
+    # Make gradient table and save to pickle
+    [gtab_file, gtab, _, _] = make_gtab(fbval_rescaled, fbvec_rescaled, vector_dir, final=False)
+
+    n_bvals = len(np.unique(bval_rescaled_mag))
+    check_val = check_multi_b(gtab, n_bvals, non_zero=False, bmag=bmag)
+    if check_val is False:
+        raise ValueError('Insufficient b-values in gradient table based on the number expected')
+
+    # Check that number of image volumes corresponds to the number of b-vector directions
+    if dwi_file:
+        volume_num = nib.load(dwi_file).shape[-1]
+
+        if not len(bvecs) == volume_num:
+            raise Exception("Expected %d total image samples but found %d total vectors", volume_num, len(bvecs))
+
+    # Check hemisphere coverage
+    [is_hemi, pole] = is_hemispherical(np.round(bvec_rescaled, 8)[~(np.round(bvec_rescaled, 8) == 0).all(1)])
+
+    # Second-level model logic
+    if is_hemi is True:
+        slm = 'linear'
+        print("%s%s%s" % ("B-vectors for this data are hemispherical at polar vertex: ", pole,
+              "To ensure adequate eddy current correction, eddy will be run using the --slm=linear option."))
+    elif len(bvecs) < 30:
+        slm = 'linear'
+        print('Fewer than 30 diffusion-weighted directions detected. To ensure adequate eddy current correction, eddy will be run using the --slm=linear option')
+    else:
+        slm = 'none'
```

Comment:
I don't think this is necessary here, please run away of uber-functions trying to solve everything.

Reply:
Will do. If I transform checkvecs into an interface (it sounds like this may be the way to go), I could just add direction count and is_hemi as outputs? 

Would the slm logic here be best as a separate function to save for later, should it just be woven into the eddy interface somehow, or should it be incorporated in some other way? If we are planning to use FSL's eddy (even in the short-term), the slm arg will be important.

Reply:
Please note before updating that I would not remove these functions, I think we want to wrap them up in Nipype interfaces.

Having these split functions makes it easier to implement unit tests (without all nipype's boilerplate).